README
This is the public release of VBPP 1.1.0.  VBPP is a Verilog preprocessor. 
It has support for most Verilog preprocessing directives and additional
directives such as:

1.  Statement generator ('generate' command in VHDL).
2.  Expression evaluation.
3.  Mathematical functions: log2, ceil, floor, round, abs, etc.
4.  Conditionals: if, switch, etc.

This software is a re-release of Hemi Thaker's VPP program which, with
version 1.3, was released under a different license.  Therefore, this
release is based on version 1.2.  There are still some features which are
not implemented yet.  The following was changed in VPP version 1.2 to
produce VBPP 1.0.

1.  ANSI-fied function prototypes.
2.  ANSI-fied library calls.
3.  Code cleanup (removed unused variables and functions).
4.  Redone makefile.
5.  Re-organized files (names).
6.  Configure script (automatic generation of Makefile).
7.  Lots of bug fixes, see CHANGELOG for a list.

The following was updated for VBPP 1.1.0:

1.  C style preprocessing with the -E option.
2.  Generate a `line directive with the -L option.
3.  Support -D option to define a macro.
4.  Support the +incdir+ option for include search directories.
5.  Macro function definition and expansion.

The following was updated for VBPP 1.0.4:

Return non-zero when parser detects error.  This could cause older scripts
to fail if it looks at the exit value.

The following was updated for VBPP 1.0.3:

1.  Fixed stack overflow problem with parser.
2.  Source available in CVS server.  The server is hosted by

	http://geda.seul.org/

    Visit http://www.flex.com/~jching for more details.

The following was updated for VBPP 1.0.2:

1.  Fixed infinite loop bug in eval_stack.
2.  Cleaned up Makefile.
3.  Added more examples for testing.

The following was updated for VBPP 1.0.1:

1.  Runs correctly on systems that does not have rint.
2.  Implemented U* functions.

The following are things that was not changed in the distribution:

1.  Features
2.  Documentation (the original README is renamed to NOTES)
3.  License

VBPP was developed with the following tools:

GNU Make 3.75
GCC 2.95.3
Bison 1.28
Flex 2.5.4

Later versions of these tools should work.  There are no gaurantees for
earlier versions.  The distribution does not contain generated files, so
Bison and Flex are required to build from source.

To install, type the following command in your favorite shell and follow
the directions given. 

./configure

This release was done with the intention of continuing to produce a freely
available Verilog preprocessor.  If you have any comments or suggestions,
please email jching@flex.com.  Note, Hemi Thaker has not endorse this branch
of his VPP software.  Thus do not email him questions and/or requests.
Please direct all comments concerning VBPP to me only.

Jimen Ching


CHANGELOG
2002-02-02  Jimen Ching  <jching@flex.com>

	* Makefile.in:
	Generate tarball with directory of vbpp-VERNAME.

	* README: Update for new release.

2002-01-31  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Detect a directive specifier ` in a name token is an error.

2002-01-30  Jimen Ching  <jching@flex.com>

	* EXAMPLES/1exout-template, EXAMPLES/x_arguments.vpp, EXAMPLES/x_macro.vpp, Makefile.in:
	New test samples for new macro function feature.

	* vpp.l: Support macro functions.

2002-01-28  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Handle long lines more gracefully, do write beyond end of buffer.

	* common.h:
	Increase the maximum line length supported.

	* vpp.y:
	Stop parsing after 100 parse errors.

	* vpp.c:
	Add -h option to print out command line options list.

	* Makefile.in:
	Generate yacc parser before lexor, since the lexor needs some yacc
	symbols.

	* create.c, vpp.l:
	Store macro definition only when it is active.

2002-01-27  Jimen Ching  <jching@flex.com>

	* vpp.y:
	Do not output comment starters.  That is done in lexor.

2002-01-26  Jimen Ching  <jching@flex.com>

	* vpp.l: Fix typo in vectornets.
	Expand `name macros when C style preprocessing is enabled.
	Store comments in MPP state, otherwise, it is outputed too early.

	* vpp.c:
	Add option to silence status message.

	* create.c:
	Do not evaluate expression for `ifdef/`ifndef directives.

2002-01-20  Jimen Ching  <jching@flex.com>

	* create.c, vpp.c, yacc_stuff.h:
	Add command line option to output `line directive and process `ifdef
	conditionals like C preprocessors.

	* vpp.l:
	nl_count was incremented too soon.

	* configure, configure.in:
	Make output of line directive a command line option instead of
	configure option.

	* EXAMPLES/1exout-template, EXAMPLES/2exout-template, vpp.l:
	Code cleanup.
	Made error/warning messages not as verbose.

2002-01-18  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Allow `keyword for INITIAL and COPY states only.  Addition of comment
	states caused processing of these directives if not restricted.

	* common.h, proto.h, vpp.c, vpp.l:
	Add support for +incdir+ and -D options.

	* EXAMPLES/1exout-template, EXAMPLES/x_cmddefine.vpp, EXAMPLES/x_hdr.vhh, EXAMPLES/x_inc.vpp, Makefile.in:
	New tests for +incdir+ and -D options.

2002-01-13  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Use the lex facilities to handle C style comments.
	Other code cleanup.

	* EXAMPLES/0exout-template, EXAMPLES/1exout-template, EXAMPLES/2exout-template, vpp.c, vpp.l, yacc_stuff.h:
	Support generating `line directive.
	Fixed bug where `include directive incorrectly tests the first include.
	Added `line directive to the beginning of parsing.
	Increased limit for include depth and `line output.

	* Makefile.in:
	Do not use symbolic link to run through examples.

	* configure, configure.in:
	Support generating `line directives.

2001-03-11  Jimen Ching  <jching@flex.com>

	* README:
	Updated for release 1.0.4.

	* CHANGELOG:
	Updated new checkin comments.

--------------------------------------------------------------------------------

2001-01-25  Jimen Ching  <jching@flex.com>

	* Makefile.in, vpp.c, vpp.y, yacc_stuff.h:
	Return non-zero when parsing error detected.

2000-11-08  Jimen Ching  <jching@flex.com>

	* Makefile.in:
	Delete EXAMPLES only if it is a symbolic link.

2000-08-20  Jimen Ching  <jching@flex.com>

	* common.h:
	More ANSI conformance.

2000-08-11  Jimen Ching  <jching@flex.com>

	* create.c:
	Handle qouted string in expression evaluation.

2000-07-29  Jimen Ching  <jching@flex.com>

	* EXAMPLES/1exout-template, EXAMPLES/x_define.vpp, Makefile.in:
	New example code to test preprocessor macro define handling.

	* create.c:
	Use new get_define_value for error detection.
	Skip 'for' loop if expression is false on entry.

	* proto.h, vpp.l:
	Changed get_define_value API to allow error detection.

2000-07-09  Jimen Ching  <jching@flex.com>

	* EXAMPLES/simple_switch.vpp, EXAMPLES/x_swfw.vpp, create.c:
	The outer active state was not taken into account when processing `endif.

--------------------------------------------------------------------------------

Thu Dec 30 21:06:21 1999 Emile van Rooyen <emile.van.rooyen@microchip.com>

	* [vpp.y]
	Prevent stack overflows.

Thu Jul  1 21:52:30 1999  Jimen Ching  <jching@flex.com>

        * [create.c]
	Fixed bug in eval_stack where `endfor is not skipped when it should be.

	* [configure.in Makefile.in]
	Code cleanup.

	* [EXAMPLES]
	Added more examples for testing.

Fri Feb 13 21:08:00 1998  Jimen Ching  <jching@flex.com>

        * [create.c]
	Added UPIPE, UNOR, UXNOR.

Mon Feb  9 21:43:43 1998  Jeffrey Ebert  <jebert@hns.com>

	* [proto.h vpp.l]
	Changed call from rint to round_int.

Sat Jan 10 11:08:39 1998  Jimen Ching  <jching@flex.com>

	* [create.c]
	Added UAND, UNAND, UNOR.


vpp.c
/* Copyright (C) 1996 Himanshu M. Thaker

This file is part of vpp.

Vpp is distributed in the hope that it will be useful,
but without any warranty.  No author or distributor
accepts responsibility to anyone for the consequences of using it
or for whether it serves any particular purpose or works at all,
unless he says so in writing.

Everyone is granted permission to copy, modify and redistribute
vpp, but only under the conditions described in the
document "vpp copying permission notice".   An exact copy
of the document is supposed to have been given to you along with
vpp so that you can know how you may redistribute it all.
It should be in a file named COPYING.  Among other things, the
copyright notice and this notice must be preserved on all copies.  */

/*
 * Program : vpp
 * Author : Himanshu M. Thaker
 * Date : Apr. 16, 1995
 * Description :
 */

#include "common.h"
#include "yacc_stuff.h"
#include "proto.h"
#include <string.h>
#include <stdio.h>

struct vpp_filename
	{
	struct vpp_filename *next;   
	char *name;
	};
struct vpp_filename *vpp_incdir;
int output_quite;

int
main(int argc, char *argv[])
	{
	int i, arglen;
	struct vpp_filename *t, *filehead, *filecurr, *idcurr;

#ifdef YYDEBUG
	yydebug = TRUE;
#endif

	/* Default to old behavior. */
	output_line_directive = FALSE;
	output_ifdef_directive = TRUE;

	filehead = filecurr = idcurr = NULL;
	for (i = 1; i < argc; ++i)
		{
		arglen = strlen(argv[i]);
		if (arglen >= 8 && !strncmp(argv[i],"+incdir+",8))
			{
			if (arglen == 8)
				{
				/* Perhaps we don't even need to notify user. */
				fprintf(stderr, "** No path for +incdir+ in argument %d.\n", i);
				}
			else
				{   
				char *lname, *tok;

				lname = (char *)NMALLOC(arglen-8+1, char);
				strcpy(lname, argv[i]+8);
				tok = strtok(lname, "+");
				while (tok)
					{
					int toklen = strlen(tok);
					t = CALLOC(1, struct vpp_filename);
					t->name = NMALLOC(toklen+1, char);
					strcpy(t->name, tok);
					if (!vpp_incdir)
						vpp_incdir = t;
					else
						idcurr->next = t;
					idcurr = t;
					tok = strtok(NULL, "+");
					}
				free(lname);
				}
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-D", 2))
			{
			if (arglen == 2)
				{
				/* Perhaps we don't even need to notify user. */
				fprintf(stderr, "** No define for -D in argument %d.\n", i);
				}
			else
				{
				char *dname, *pnt;

				dname = NMALLOC(arglen+8, char);
				sprintf(dname, "`define %s", argv[i]+2);
				pnt = dname+8;
				while (*pnt)
					{
					if (*pnt == '=')
						{
						*pnt = ' ';
						break;
						}
					pnt++;
					}
				store_define(dname);
				free(dname);
				}
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-E", 2))
			{
			/* Enable C style preprocessing of ifdef and friends. */
			output_ifdef_directive = FALSE;
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-L", 2))
			{
			/* Write out `line directives for debugging. */
			output_line_directive = TRUE;
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-h", 2))
			{
			printf("syntax: %s [options] filename\n", argv[0]);
			printf("options:\n");
			printf("  +incdir+...+ Search directory for `include files.\n");
			printf("  -D           Define macro.\n");
			printf("  -E           Perform C style preprocessing.\n");
			printf("  -L           Output `line directive.\n");
			printf("  -h           Print help message and exit.\n");
			printf("  -q           Suppress status message.\n");
			return 0;
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-q", 2))
			{
			/* Quite, take out unnecessary output. */
			output_quite = TRUE;
			}
		else /* filename only.. */
			{
			t = CALLOC(1,struct vpp_filename);
			t->name = NMALLOC(strlen(argv[i])+1, char);
			strcpy(t->name, argv[i]);
			if (!filehead)
				filehead = t;
			else
				filecurr->next = t;
			filecurr = t;
			}
        }

	/*
	 * check to see if the file can be opened for reading...
	 */
	if (!filehead)
		{
		yyin = stdin;
		current_file = NMALLOC(strlen("<stdin>")+1, char);
		strcpy(current_file, "<stdin>");

		nl_count = 1;
		if (!output_quite)
			fprintf(stderr, "Preprocessing '%s'\n", current_file);
		do_comment_count(TRUE, nl_count);
		yyparse();
		fclose(yyin);
		free(current_file);
		}
	else
		{
		filecurr = filehead;
		while (filecurr)
			{
			yyin = fopen(filecurr->name, "r");
      		if (yyin == NULL)
				{
        		fprintf(stderr, "** %s: open failed\n", filecurr->name);
				++yyerror_count;
				}
			else
				{
      			current_file = NMALLOC(strlen(filecurr->name)+1, char);
      			strcpy(current_file, filecurr->name);

				nl_count = 1;
				if (!output_quite) {
					fprintf(stderr, "vpp 1.2.0, see 'https://github.com/balanx/vbpp'\n");
					fprintf(stderr, "Preprocessing '%s'\n", current_file);
				}
				do_comment_count(TRUE, nl_count);
				yyparse();
				fclose(yyin);
				free(current_file);
				}

			t = filecurr;
			filecurr = filecurr->next;
			free(t->name);
			free(t);
			}

		filehead = NULL;
		}

	return yyerror_count;
	}

FILE *
incdir_fopen(char *name)
	{
	int len;
	struct vpp_filename *l;  
	char *path;
	FILE *fp = NULL;

	len = strlen(name);
	l = vpp_incdir;
	while (l)
        {
        path = (char *)malloc(strlen(l->name) + 1 + len + 1);
        sprintf(path, "%s/%s", l->name, name);
        fp = fopen(path, "r");
        free(path);
        if (fp)
			break;
		l = l->next;
        }

	if (!fp)
		fp = fopen(name, "r");
	return(fp);
	}


common.h
/* Copyright (C) 1996 Himanshu M. Thaker

This file is part of vpp.

Vpp is distributed in the hope that it will be useful,
but without any warranty.  No author or distributor
accepts responsibility to anyone for the consequences of using it
or for whether it serves any particular purpose or works at all,
unless he says so in writing.

Everyone is granted permission to copy, modify and redistribute
vpp, but only under the conditions described in the
document "vpp copying permission notice".   An exact copy
of the document is supposed to have been given to you along with
vpp so that you can know how you may redistribute it all.
It should be in a file named COPYING.  Among other things, the
copyright notice and this notice must be preserved on all copies.  */

#ifndef VPP_COMMON_H
#define VPP_COMMON_H

/*
 * first, the include files...
 */
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

/*
 * then, some globally accessible defines...
 */
#define MAXSTR 256
#define MAXNAME 16384
#define BASE 89
#define	FALSE	0
#define	TRUE	1
#define MALLOC(type)     ((type *) malloc(sizeof(type)))
#define NMALLOC(n,type)  ((type *) malloc((n)*sizeof(type)))
#define CALLOC(n,type)   ((type *) calloc((n),sizeof(type)))

#endif /* VPP_COMMON_H */


yacc_stuff.h
/* Copyright (C) 1996 Himanshu M. Thaker

This file is part of vpp.

Vpp is distributed in the hope that it will be useful,
but without any warranty.  No author or distributor
accepts responsibility to anyone for the consequences of using it
or for whether it serves any particular purpose or works at all,
unless he says so in writing.

Everyone is granted permission to copy, modify and redistribute
vpp, but only under the conditions described in the
document "vpp copying permission notice".   An exact copy
of the document is supposed to have been given to you along with
vpp so that you can know how you may redistribute it all.
It should be in a file named COPYING.  Among other things, the
copyright notice and this notice must be preserved on all copies.  */

#ifndef __YACC_STUFF_H__
#define __YACC_STUFF_H__

extern FILE *yyin;
extern char *yytext;
extern int yyerror_count;
extern void yyerror(char *str); /* Our version. */
extern int yywrap(void); /* Our version. */
extern int yylex(void); /* Lexical analyzer function. */
extern int yyparse(void); /* Parser function. */

/* vpp.l */
extern int nl_count;
extern char old_line[MAXNAME];
extern char current_line[MAXNAME];
extern char *current_file;
extern int output_line_directive;
extern int output_ifdef_directive;

typedef void ll_generic;

#define DELTA 0.0001

#define INTEGER 0
#define REAL 1
#define STRING 3
#define QUOTED_STRING 98
#define NMBR 99

#define ADD 4
#define SUB 5
#define MULT 6
#define DIV 7
#define MOD 8
#define EQ 9
#define NEQ 10
#define CASEEQ 11
#define CASENEQ 12
#define AND 13
#define OR 14
#define LT 15
#define LEQ 16
#define GT 17
#define GEQ 18
#define BITAND 19
#define BITOR 20
#define BITXOR 21
#define BITXNOR 22
#define SHIFTR 23
#define SHIFTL 24

#define UPLUS 25
#define UMINUS 26
#define UNOT 27
#define UTILDA 28
#define UAND 29
#define UNAND 30
#define UNOR 31
#define UPIPE 32
#define UCARAT 33
#define UXNOR 34

#define TRINARY 35

#define ID_SINGLE_EXPR 36
#define ID_DOUBLE_EXPR 37
#define CONCATENATION 38
#define MULTIPLE_CONCATENATION 39
#define FUNCTION 40

#define MINTYPMAX 41
#define POWER 42
#define LOG2 43
#define ROUND 44
#define CEIL 45
#define FLOOR 46
#define MAX 47
#define MIN 48
#define ABS 49
#define STOI 50
#define ITOS 51
#define SUBSTR 52
#define SYSTEM 53
#define EVEN 54
#define ODD 55

#define IDENT_REF_IDENT 0
#define IDENT_REF_ONE_EXPR 1
#define IDENT_REF_TWO_EXPR 2
#define IDENT_REF_CONCAT 3

#define STMT_BLOCKING 0
#define STMT_NON_BLOCKING 1
#define STMT_IF 2
#define STMT_CASE 3
#define STMT_CASEZ 4
#define STMT_CASEX 5
#define STMT_FOREVER 6
#define STMT_REPEAT 7
#define STMT_WHILE 8
#define STMT_FOR 9
#define STMT_WAIT 10
#define STMT_EVENT_TRIGGER 11

typedef struct expr {
    short ident;
    union {
	int val_int;
	double val_real;
	char *strg;
    } u;
    struct expr *ep1;
    struct expr *ep2;
    struct expr *ep3;
    struct expr_list *elp;
    struct var_values *vvp;
    int precision;
} expr;

typedef struct expr_list {
    expr *ep;
    struct expr_list *next;
} expr_list;

typedef struct t_ident_ref {
    short ident;
    char *strg;
    struct expr *ep1;
    struct expr *ep2;
} t_ident_ref;

typedef struct stmt {
    short ident;
    struct expr *ep;
    struct stmt *sp1;
    struct stmt *sp2;
    struct t_ident_ref *lp;
} stmt;

typedef struct t_module {
    char *name;
    struct t_port_list *port_list;
    struct t_mod_item *mod_item;
} t_module;

typedef struct varp {
    char *name;
    expr *ep;
    struct var_values *vvp;
    char *strg;
} varp;

typedef struct var_list {
    varp *vp;
    struct var_list *next;
} var_list;

#define EV_EXP_LIST 0
#define F_OR_S_LIST 1
#define GENERIC_LIST 2
#define PRESET_LIST 3
#define RESET_LIST 4
#define EXCLUDE_LIST 5
#define NAMES_LIST 6
#define PARAM_LIST 7

#define CODE_POINTER 8
#define FOR_STRUCTURE 9
#define IF_STRUCTURE 10
#define IFDEF_STRUCTURE 11
#define IFNDEF_STRUCTURE 12
#define WHILE_STRUCTURE 13
#define ENDIF_STRUCTURE 14
#define ELSE_STRUCTURE 15
#define COLLAPSED_STRUCTURE 16
#define CODE_POINTER2 17
#define ENDFOR_STRUCTURE 18
#define LET_STRUCTURE 19
#define ENDWHILE_STRUCTURE 20
#define SWITCH_STRUCTURE 21
#define CASE_STRUCTURE 22
#define DEFAULT_STRUCTURE 23
#define ENDSWITCH_STRUCTURE 24
#define DEFINE_STRUCTURE 25
#define BREAKSW_STRUCTURE 26

typedef struct gen_ll {
    int ll_type;
    int line_count;
    ll_generic *item;
    struct gen_ll *prev;
    struct gen_ll *next;
} gen_ll;

/*
 * the following stucture points to the head and tail of a
 * generic linked list.
 */
typedef struct gen_ll_mgr {
    struct gen_ll *head;
    struct gen_ll *tail;
} gen_ll_mgr;

/*
 * for structure...
 */
typedef struct for_struct {
    varp *vp1;
    expr *ep_eval;
    char *name2;
    expr *ep2;
    int eval_first;
} for_struct;

/*
 * while structure...
 */
typedef struct while_struct {
    expr *ep;
    int eval_first;
} while_struct;

/*
 * switch structure...
 */
typedef struct switch_struct {
    expr *ep;
    int done_a_case;
    int status_on_entry;
    int ignore_more_matches;
    int value;
} switch_struct;

typedef struct var_values {
    char *name;
    int value;
    double dval;
    int vtype;
    char *strg;
    struct var_values *next;
} var_values;

typedef struct loop_stack
{
    int ll_type;
    int active;
    int region_active;
    gen_ll *mate;
    expr *ep;
    struct loop_stack *next;
} loop_stack;

#endif /* __YACC_STUFF_H__ */
